TypeError when writing table column with mixed string / NaN

When writing a SpatialData object which has a column containing string values or empty values, an error is raised.

Empty values are a very common use-case:
- When two regions have fundamentally different annotations, there is no possible value to fill in for the other region. This is often represented as None or NaN. For example region "cells" could have column "cell_diameter" but region "well" cannot have "cell_diameter".
- When concatenating tables where some columns existed only in either one of the source tables. Due to the single-table design, SpatialData needs to fill up these columns, which it does with NaN. 
  This can also be a huge problem for integer columns which have no NaN. Pandas silently changes the dtype to float. Then whoever wants to read values from the supposed integer column must precautionarily convert to int to avoid follow up errors.

**Example**

```python
from anndata import AnnData
from spatialdata import SpatialData
from spatialdata.models import TableModel
import numpy as np
import pandas as pd
import pytest
import shutil


@pytest.fixture
def sdata_with_nan_in_obs():
    table = TableModel.parse(
        AnnData(
            obs=pd.DataFrame(
                {
                    "region": ["region1", "region2"],
                    "instance": [0, 0],
                    "column_only_region1": ["string", np.nan],
                },
                index=pd.RangeIndex(0, 2).astype(str)
            ).astype({"region": "category"}),
        ),
        region_key="region",
        instance_key="instance",
        region=["region1", "region2"],
    )
    return SpatialData(table=table)


def test_sdata_with_nan_in_obs(sdata_with_nan_in_obs, tmp_path):
    sdata = sdata_with_nan_in_obs
    # Remove any pre-existing files for being able to repeat the test
    shutil.rmtree(tmp_path / "sdata_with_nan_in_obs.zarr", ignore_errors=True)
    # Raises "TypeError: expected unicode string, found nan"
    sdata.write(tmp_path / "sdata_with_nan_in_obs.zarr")
```

**Backtrace**

```
../../../../../spatialdata/src/spatialdata/_core/spatialdata.py:1052: in write
    raise e
../../../../../spatialdata/src/spatialdata/_core/spatialdata.py:1048: in write
    write_table(table=self.table, group=elem_group, name="table")
../../../../../spatialdata/src/spatialdata/_io/io_table.py:20: in write_table
    write_adata(group, name, table)  # creates group[name]
 …/site-packages/anndata/_io/specs/registry.py:353: in write_elem
    Writer(_REGISTRY).write_elem(store, k, elem, dataset_kwargs=dataset_kwargs)
 …/site-packages/anndata/_io/utils.py:248: in func_wrapper
    re_raise_error(e, elem, key)
 …/site-packages/anndata/_io/utils.py:246: in func_wrapper
    return func(*args, **kwargs)
 …/site-packages/anndata/_io/specs/registry.py:311: in write_elem
    return write_func(store, k, elem, dataset_kwargs=dataset_kwargs)
 …/site-packages/anndata/_io/specs/registry.py:52: in wrapper
    result = func(g, k, *args, **kwargs)
 …/site-packages/anndata/_io/specs/methods.py:220: in write_anndata
    _writer.write_elem(g, "obs", adata.obs, dataset_kwargs=dataset_kwargs)
 …/site-packages/anndata/_io/utils.py:248: in func_wrapper
    re_raise_error(e, elem, key)
 …/site-packages/anndata/_io/utils.py:246: in func_wrapper
    return func(*args, **kwargs)
 …/site-packages/anndata/_io/specs/registry.py:311: in write_elem
    return write_func(store, k, elem, dataset_kwargs=dataset_kwargs)
 …/site-packages/anndata/_io/specs/registry.py:52: in wrapper
    result = func(g, k, *args, **kwargs)
 …/site-packages/anndata/_io/specs/methods.py:579: in write_dataframe
    _writer.write_elem(
 …/site-packages/anndata/_io/utils.py:248: in func_wrapper
    re_raise_error(e, elem, key)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

e = TypeError('expected unicode string, found nan')
elem = <zarr.hierarchy.Group '/table/table/obs'>, key = 'column_only_region1'

    def re_raise_error(e, elem, key):
        if "Above error raised while writing key" in format(e):
            raise
        else:
            parent = _get_parent(elem)
>           raise type(e)(
                f"{e}\n\n"
                f"Above error raised while writing key {key!r} of {type(elem)} "
                f"to {parent}"
            ) from e
E           TypeError: expected unicode string, found nan
E           
E           Above error raised while writing key 'column_only_region1' of <class 'zarr.hierarchy.Group'> to <zarr.storage.FSStore object at 0x7fba5b08cd30>
```

This is not directly caused by `anndata`, and interestingly, anndata modifies the table so that the column "column_only_region1" gets dtype `"category"` instead of `object`, and the error does not occur anymore:
```python
def test_adata_with_nan_in_obs(sdata_with_nan_in_obs, tmp_path):
    adata = sdata_with_nan_in_obs.table
    # AnnData does not cause an error
    adata.write_zarr(tmp_path / "adata_with_nan_in_obs.zarr")
    # and modifies the table so that SpatialData does not raise the error anymore:
    shutil.rmtree(tmp_path / "sdata_with_nan_in_obs.zarr", ignore_errors=True)
    sdata_with_nan_in_obs.write(tmp_path / "sdata_with_nan_in_obs.zarr")
    # No error raised
```

--- a/src/spatialdata/models/models.py
+++ b/src/spatialdata/models/models.py
@@ -1061,6 +1061,21 @@ class TableModel:
         if len(set(expected_regions).symmetric_difference(set(found_regions))) > 0:
             raise ValueError(f"Regions in the AnnData object and `{attr[self.REGION_KEY_KEY]}` do not match.")
 
+        # Warning for object/string columns with NaN in region_key or instance_key
+        instance_key = attr[self.INSTANCE_KEY]
+        region_key = attr[self.REGION_KEY_KEY]
+        for key_name, key_value in [("region_key", region_key), ("instance_key", instance_key)]:
+            if key_value in data.obs:
+                col = data.obs[key_value]
+                col_dtype = col.dtype
+                if (col_dtype == "object" or pd.api.types.is_string_dtype(col_dtype)) and col.isna().any():
+                    logger.warning(
+                        f"The {key_name} column '{key_value}' is of {col_dtype} type and contains NaN values. "
+                        "After writing and reading with AnnData, NaN values may (depending on the AnnData version) "
+                        "be converted to strings. This may cause issues when matching instances across read/write "
+                        "cycles."
+                    )
+
     def validate(
         self,
         data: AnnData,

--- a/tests/io/test_readwrite.py
+++ b/tests/io/test_readwrite.py
@@ -1065,3 +1065,45 @@ def test_read_sdata(tmp_path: Path, points: SpatialData) -> None:
     assert_spatial_data_objects_are_identical(sdata_from_path, sdata_from_str)
     assert_spatial_data_objects_are_identical(sdata_from_path, sdata_from_upath)
     assert_spatial_data_objects_are_identical(sdata_from_path, sdata_from_zarr_group)
+
+
+def test_sdata_with_nan_in_obs() -> None:
+    """Test writing SpatialData with mixed string/NaN values in obs works correctly.
+
+    Regression test for https://github.com/scverse/spatialdata/issues/399
+    Previously this raised TypeError: expected unicode string, found nan.
+    Now the write succeeds, though NaN values in object-dtype columns are
+    converted to the string "nan" after round-trip.
+    """
+    from spatialdata.models import TableModel
+
+    table = TableModel.parse(
+        AnnData(
+            obs=pd.DataFrame(
+                {
+                    "region": ["region1", "region2"],
+                    "instance": [0, 0],
+                    "column_only_region1": ["string", np.nan],
+                    "column_only_region2": [np.nan, 3],
+                }
+            )
+        ),
+        region_key="region",
+        instance_key="instance",
+        region=["region1", "region2"],
+    )
+    sdata = SpatialData(tables={"table": table})
+    assert sdata["table"].obs["column_only_region1"].iloc[1] is np.nan
+    assert np.isnan(sdata["table"].obs["column_only_region2"].iloc[0])
+
+    with tempfile.TemporaryDirectory() as tmpdir:
+        path = os.path.join(tmpdir, "data.zarr")
+        sdata.write(path)
+
+        sdata2 = SpatialData.read(path)
+        assert "column_only_region1" in sdata2["table"].obs.columns
+        assert sdata2["table"].obs["column_only_region1"].iloc[0] == "string"
+        assert sdata2["table"].obs["column_only_region2"].iloc[1] == 3
+        # After round-trip, NaN in object-dtype column becomes string "nan"
+        assert sdata2["table"].obs["column_only_region1"].iloc[1] == "nan"
+        assert np.isnan(sdata2["table"].obs["column_only_region2"].iloc[0])

